<a href="https://colab.research.google.com/github/nsk20/Financial-Market-Analysis/blob/main/Gold_to_INR.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import pandas as pd

# Task
Load and combine 'Nifty 50 Historical Data.csv' and 'Nifty 50 Historical Data2.csv' into `nifty_df`; 'S&P 500 Historical Data.csv' and 'S&P 500 Historical Data2.csv' into `sp500_df`; 'Gold Futures Historical Data.csv' and 'Gold Futures Historical Data2.csv' into `gold_df`; and 'USD_INR Historical Data.csv' and 'USD_INR Historical Data2.csv' into `usd_inr_df`, then confirm creation and summarize the data.

## Load and combine Nifty 50 data

### Subtask:
Load 'Nifty 50 Historical Data.csv' and 'Nifty 50 Historical Data2.csv' into pandas DataFrames and combine them into a single DataFrame named `nifty_df`.


**Reasoning**:
Load the two Nifty 50 CSV files into separate pandas DataFrames, then concatenate them into a single DataFrame named `nifty_df` and display its head.



In [ ]:
nifty_df_part1 = pd.read_csv('/content/Nifty 50 Historical Data.csv')
nifty_df_part2 = pd.read_csv('/content/Nifty 50 Historical Data2.csv')
nifty_df = pd.concat([nifty_df_part1, nifty_df_part2], ignore_index=True)
nifty_df.head()

Date      Price       Open       High        Low     Vol. Change %
0  02/25/2025  22,547.55  22,516.45  22,625.30  22,513.90  250.28M   -0.03%
1  02/24/2025  22,553.35  22,609.35  22,668.05  22,518.80  214.32M   -1.06%
2  02/21/2025  22,795.90  22,857.20  22,921.00  22,720.30  242.75M   -0.51%
3  02/20/2025  22,913.15  22,821.10  22,923.85  22,812.75  240.84M   -0.09%
4  02/19/2025  22,932.90  22,847.25  23,049.95  22,814.85  207.00M   -0.05%

## Load and combine S&P 500 data

### Subtask:
Load 'S&P 500 Historical Data.csv' and 'S&P 500 Historical Data2.csv' into pandas DataFrames and combine them into a single DataFrame named `sp500_df`.


**Reasoning**:
Load the two S&P 500 CSV files into separate pandas DataFrames, concatenate them into a single DataFrame named `sp500_df`, and then display its head to verify the merge.



In [ ]:
sp500_df_part1 = pd.read_csv('/content/S&P 500 Historical Data.csv')
sp500_df_part2 = pd.read_csv('/content/S&P 500 Historical Data2.csv')
sp500_df = pd.concat([sp500_df_part1, sp500_df_part2], ignore_index=True)
sp500_df.head()

Date     Price      Open      High       Low  Vol. Change %
0  11/12/2024  5,983.99  6,003.60  6,009.92  5,960.08   NaN   -0.29%
1  11/11/2024  6,001.35  6,008.86  6,017.31  5,986.69   NaN    0.10%
2  11/08/2024  5,995.54  5,976.76  6,012.45  5,976.76   NaN    0.38%
3  11/07/2024  5,973.10  5,947.21  5,983.84  5,947.21   NaN    0.74%
4  11/06/2024  5,929.04  5,864.89  5,936.14  5,864.89   NaN    2.53%

## Load and combine Gold Futures data

### Subtask:
Load 'Gold Futures Historical Data.csv' and 'Gold Futures Historical Data2.csv' into pandas DataFrames and combine them into a single DataFrame named `gold_df`.

**Reasoning**:
Load the two Gold Futures CSV files into separate pandas DataFrames, then concatenate them into a single DataFrame named `gold_df` and display its head to verify the merge.



In [ ]:
gold_df_part1 = pd.read_csv('/content/Gold Futures Historical Data.csv')
gold_df_part2 = pd.read_csv('/content/Gold Futures Historical Data2.csv')
gold_df = pd.concat([gold_df_part1, gold_df_part2], ignore_index=True)
gold_df.head()

Date     Price      Open      High       Low   Vol. Change %
0  08/26/2024  2,520.20  2,512.60  2,527.70  2,509.30  0.66K    0.36%
1  08/23/2024  2,511.20  2,486.20  2,518.70  2,486.10  0.75K    1.17%
2  08/22/2024  2,482.10  2,515.60  2,516.00  2,472.60  0.74K   -1.22%
3  08/21/2024  2,512.80  2,517.40  2,522.00  2,495.00  0.60K   -0.13%
4  08/20/2024  2,516.10  2,507.70  2,535.40  2,502.10  0.66K    0.37%

## Load and combine USD/INR data

### Subtask:
Load 'USD_INR Historical Data.csv' and 'USD_INR Historical Data2.csv' into pandas DataFrames and combine them into a single DataFrame named `usd_inr_df`.

**Reasoning**:
Load the two USD/INR CSV files into separate pandas DataFrames, then concatenate them into a single DataFrame named `usd_inr_df` and display its head to verify the merge.



In [ ]:
usd_inr_df_part1 = pd.read_csv('/content/USD_INR Historical Data.csv')
usd_inr_df_part2 = pd.read_csv('/content/USD_INR Historical Data2.csv')
usd_inr_df = pd.concat([usd_inr_df_part1, usd_inr_df_part2], ignore_index=True)
usd_inr_df.head()

Date   Price    Open    High     Low  Vol. Change %
0  03/01/2024  82.841  82.926  82.933  82.836   NaN   -0.07%
1  02/29/2024  82.900  82.910  82.957  82.864   NaN    0.00%
2  02/28/2024  82.900  82.924  82.967  82.863   NaN    0.03%
3  02/27/2024  82.877  82.901  82.921  82.841   NaN    0.04%
4  02/26/2024  82.846  82.873  82.914  82.851   NaN   -0.02%

In [ ]:
nifty_df['Date'] = pd.to_datetime(nifty_df['Date'], format='%m/%d/%Y')
print(f"Nifty 50 Dates: Min Date = {nifty_df['Date'].min()}, Max Date = {nifty_df['Date'].max()}")
sp500_df['Date'] = pd.to_datetime(sp500_df['Date'], format='%m/%d/%Y')
print(f"S&P 500 Dates: Min Date = {sp500_df['Date'].min()}, Max Date = {sp500_df['Date'].max()}")
gold_df['Date'] = pd.to_datetime(gold_df['Date'], format='%m/%d/%Y')
print(f"Gold Futures Dates: Min Date = {gold_df['Date'].min()}, Max Date = {gold_df['Date'].max()}")
usd_inr_df['Date'] = pd.to_datetime(usd_inr_df['Date'], format='%m/%d/%Y')
print(f"USD/INR Dates: Min Date = {usd_inr_df['Date'].min()}, Max Date = {usd_inr_df['Date'].max()}")

Nifty 50 Dates: Min Date = 2005-01-03 00:00:00, Max Date = 2025-12-09 00:00:00
S&P 500 Dates: Min Date = 2005-01-03 00:00:00, Max Date = 2025-12-09 00:00:00
Gold Futures Dates: Min Date = 2005-01-03 00:00:00, Max Date = 2025-12-09 00:00:00
USD/INR Dates: Min Date = 2005-01-03 00:00:00, Max Date = 2025-12-09 00:00:00


In [ ]:
nifty_df = nifty_df.drop(columns=['Vol.', 'Change %'])
nifty_df.head()

Date      Price       Open       High        Low
0 2025-02-25  22,547.55  22,516.45  22,625.30  22,513.90
1 2025-02-24  22,553.35  22,609.35  22,668.05  22,518.80
2 2025-02-21  22,795.90  22,857.20  22,921.00  22,720.30
3 2025-02-20  22,913.15  22,821.10  22,923.85  22,812.75
4 2025-02-19  22,932.90  22,847.25  23,049.95  22,814.85

In [ ]:
sp500_df = sp500_df.drop(columns=['Vol.', 'Change %'])
sp500_df.head()

Date     Price      Open      High       Low
0 2024-11-12  5,983.99  6,003.60  6,009.92  5,960.08
1 2024-11-11  6,001.35  6,008.86  6,017.31  5,986.69
2 2024-11-08  5,995.54  5,976.76  6,012.45  5,976.76
3 2024-11-07  5,973.10  5,947.21  5,983.84  5,947.21
4 2024-11-06  5,929.04  5,864.89  5,936.14  5,864.89

In [ ]:
gold_df = gold_df.drop(columns=['Vol.', 'Change %'])
gold_df.head()

Date     Price      Open      High       Low
0 2024-08-26  2,520.20  2,512.60  2,527.70  2,509.30
1 2024-08-23  2,511.20  2,486.20  2,518.70  2,486.10
2 2024-08-22  2,482.10  2,515.60  2,516.00  2,472.60
3 2024-08-21  2,512.80  2,517.40  2,522.00  2,495.00
4 2024-08-20  2,516.10  2,507.70  2,535.40  2,502.10

In [ ]:
usd_inr_df = usd_inr_df.drop(columns=['Vol.', 'Change %'])
usd_inr_df.head()

Date   Price    Open    High     Low
0 2024-03-01  82.841  82.926  82.933  82.836
1 2024-02-29  82.900  82.910  82.957  82.864
2 2024-02-28  82.900  82.924  82.967  82.863
3 2024-02-27  82.877  82.901  82.921  82.841
4 2024-02-26  82.846  82.873  82.914  82.851

# Task
Convert the 'Price' column in `nifty_df` and `usd_inr_df` to a numeric (float) data type by removing commas, prepare `usd_inr_df` by setting the 'Date' column as its index, sorting it, reindexing to cover all dates within its range, and interpolating any missing 'Price' values linearly. Then, merge the processed `usd_inr_df` into `nifty_df` on the 'Date' column, and calculate a new `price_in_usd` column in `nifty_df` by dividing the Nifty 'Price' by the corresponding USD/INR exchange rate. Display the head of `nifty_df` to confirm the successful creation and accuracy of the new `price_in_usd` column.

## Convert Price columns to numeric

### Subtask:
Convert the 'Price' column in `nifty_df` and `usd_inr_df` from string to numeric (float) by removing commas.


**Reasoning**:
To convert the 'Price' column in both `nifty_df` and `usd_inr_df` to a numeric type, I need to remove commas from the string values and then cast them to float.



In [ ]:
nifty_df['Price'] = nifty_df['Price'].astype(str).str.replace(',', '', regex=False).astype(float)
usd_inr_df['Price'] = usd_inr_df['Price'].astype(float)

print("Converted 'Price' column in nifty_df to numeric:")
print(nifty_df['Price'].head())
print("\nConverted 'Price' column in usd_inr_df to numeric:")
print(usd_inr_df['Price'].head())

Converted 'Price' column in nifty_df to numeric:
0    22547.55
1    22553.35
2    22795.90
3    22913.15
4    22932.90
Name: Price, dtype: float64

Converted 'Price' column in usd_inr_df to numeric:
0    82.841
1    82.900
2    82.900
3    82.877
4    82.846
Name: Price, dtype: float64


## Prepare USD/INR for merge and interpolate

### Subtask:
Set the 'Date' column of `usd_inr_df` as the index, sort it, and then reindex it to cover all dates within its range. Interpolate any missing 'Price' values using linear interpolation to approximate rates for dates without direct matches.


**Reasoning**:
To prepare `usd_inr_df` for merging, I will set the 'Date' column as the index, sort it, reindex to a full date range, and then linearly interpolate any missing 'Price' values.



In [ ]:
usd_inr_df = usd_inr_df.set_index('Date')
usd_inr_df = usd_inr_df.sort_index()

date_range = pd.date_range(start=usd_inr_df.index.min(), end=usd_inr_df.index.max())
usd_inr_df = usd_inr_df.reindex(date_range)
usd_inr_df['Price'] = usd_inr_df['Price'].interpolate(method='linear')

print("Prepared USD/INR DataFrame:")
print(usd_inr_df.head())
print(usd_inr_df.tail())

Prepared USD/INR DataFrame:
             Price    Open    High     Low
2005-01-03  43.400  43.420  43.710  43.350
2005-01-04  43.495  43.335  43.615  43.315
2005-01-05  43.530  43.400  43.870  43.360
2005-01-06  43.875  43.695  43.985  43.665
2005-01-07  43.750  43.880  43.985  43.660
                Price    Open    High     Low
2025-12-05  89.945000  89.900  90.076  89.700
2025-12-06  90.006667     NaN     NaN     NaN
2025-12-07  90.068333     NaN     NaN     NaN
2025-12-08  90.130000  89.962  90.267  89.962
2025-12-09  90.023000  90.110  90.309  89.850


**Reasoning**:
Now that `usd_inr_df` is prepared with a complete date range and interpolated 'Price' values, I will merge it with `nifty_df` on the 'Date' column. After the merge, I will calculate the `price_in_usd` column in `nifty_df` by dividing the 'Price' from `nifty_df` by the corresponding 'Price' (exchange rate) from the merged `usd_inr_df`.



In [ ]:
nifty_df = nifty_df.set_index('Date')
nifty_df = nifty_df.sort_index()

nifty_df = nifty_df.merge(usd_inr_df[['Price']], left_index=True, right_index=True, how='left', suffixes=('_nifty', '_usd_inr'))
nifty_df['price_in_usd'] = nifty_df['Price_nifty'] / nifty_df['Price_usd_inr']
nifty_df = nifty_df.reset_index()

print("Merged nifty_df with USD/INR data and calculated price_in_usd:")
print(nifty_df.head())

Merged nifty_df with USD/INR data and calculated price_in_usd:
        Date  Price_nifty      Open      High       Low  Price_usd_inr  \
0 2005-01-03      2115.00  2,080.00  2,118.60  2,080.00         43.400   
1 2005-01-04      2103.75  2,116.95  2,120.15  2,100.55         43.495   
2 2005-01-05      2032.20  2,103.75  2,105.10  1,990.15         43.530   
3 2005-01-06      1998.35  2,031.55  2,035.65  1,984.25         43.875   
4 2005-01-07      2015.50  1,998.25  2,021.45  1,992.55         43.750   

   price_in_usd  
0     48.732719  
1     48.367628  
2     46.685045  
3     45.546439  
4     46.068571  


In [ ]:
nifty_df1 = nifty_df.drop(columns=['Price_nifty', 'Price_usd_inr'])
nifty_df1.rename(columns={'price_in_usd': 'Price'}, inplace=True)
nifty_df1['Ticker'] = 'Nifty-50'
nifty_df1.head()

Date      Open      High       Low      Price    Ticker
0 2005-01-03  2,080.00  2,118.60  2,080.00  48.732719  Nifty-50
1 2005-01-04  2,116.95  2,120.15  2,100.55  48.367628  Nifty-50
2 2005-01-05  2,103.75  2,105.10  1,990.15  46.685045  Nifty-50
3 2005-01-06  2,031.55  2,035.65  1,984.25  45.546439  Nifty-50
4 2005-01-07  1,998.25  2,021.45  1,992.55  46.068571  Nifty-50

In [ ]:
gold_df['Price'] = gold_df['Price'].astype(str).str.replace(',', '', regex=False).astype(float) / 31.1035
gold_df['High'] = gold_df['High'].astype(str).str.replace(',', '', regex=False).astype(float) / 31.1035
gold_df['Low'] = gold_df['Low'].astype(str).str.replace(',', '', regex=False).astype(float) / 31.1035
gold_df['Open'] = gold_df['Open'].astype(str).str.replace(',', '', regex=False).astype(float) / 31.1035
print("Gold Futures Price converted from Troy Ounce-to-USD to Gram-to-USD:")
display(gold_df.head())

Gold Futures Price converted from Troy Ounce-to-USD to Gram-to-USD:


Date      Price       Open       High        Low
0 2024-08-26  81.026251  80.781906  81.267381  80.675808
1 2024-08-23  80.736895  79.933126  80.978025  79.929911
2 2024-08-22  79.801309  80.878358  80.891218  79.495877
3 2024-08-21  80.788336  80.936229  81.084122  80.216053
4 2024-08-20  80.894433  80.624367  81.514942  80.444323

In [ ]:
gold_df[gold_df['Date'] == gold_df['Date'].max()]

Date       Price        Open        High         Low
5000 2025-12-09  136.229042  135.796615  136.675937  134.975164

In [ ]:
sp500_df['Price'] = sp500_df['Price'].astype(str).str.replace(',', '', regex=False).astype(float)
sp500_df['Open'] = sp500_df['Open'].astype(str).str.replace(',', '', regex=False).astype(float)
sp500_df['High'] = sp500_df['High'].astype(str).str.replace(',', '', regex=False).astype(float)
sp500_df['Low'] = sp500_df['Low'].astype(str).str.replace(',', '', regex=False).astype(float)
print("Converted 'Price', 'Open', 'High', 'Low' columns in sp500_df to numeric:")
display(sp500_df.head())

Converted 'Price', 'Open', 'High', 'Low' columns in sp500_df to numeric:


Date    Price     Open     High      Low
0 2024-11-12  5983.99  6003.60  6009.92  5960.08
1 2024-11-11  6001.35  6008.86  6017.31  5986.69
2 2024-11-08  5995.54  5976.76  6012.45  5976.76
3 2024-11-07  5973.10  5947.21  5983.84  5947.21
4 2024-11-06  5929.04  5864.89  5936.14  5864.89

In [ ]:
usd_inr_df['Open'] = usd_inr_df['Open'].astype(float)
usd_inr_df['High'] = usd_inr_df['High'].astype(float)
usd_inr_df['Low'] = usd_inr_df['Low'].astype(float)
print("Converted 'Open', 'High', 'Low' columns in usd_inr_df to numeric:")
display(usd_inr_df.head())

Converted 'Open', 'High', 'Low' columns in usd_inr_df to numeric:


Price    Open    High     Low
2005-01-03  43.400  43.420  43.710  43.350
2005-01-04  43.495  43.335  43.615  43.315
2005-01-05  43.530  43.400  43.870  43.360
2005-01-06  43.875  43.695  43.985  43.665
2005-01-07  43.750  43.880  43.985  43.660

In [ ]:
gold_df['Ticker'] = 'Gold'
gold_df.head()

Date      Price       Open       High        Low Ticker
0 2024-08-26  81.026251  80.781906  81.267381  80.675808   Gold
1 2024-08-23  80.736895  79.933126  80.978025  79.929911   Gold
2 2024-08-22  79.801309  80.878358  80.891218  79.495877   Gold
3 2024-08-21  80.788336  80.936229  81.084122  80.216053   Gold
4 2024-08-20  80.894433  80.624367  81.514942  80.444323   Gold

In [ ]:
sp500_df['Ticker'] = 'S&P500'
sp500_df.head()

Date    Price     Open     High      Low  Ticker
0 2024-11-12  5983.99  6003.60  6009.92  5960.08  S&P500
1 2024-11-11  6001.35  6008.86  6017.31  5986.69  S&P500
2 2024-11-08  5995.54  5976.76  6012.45  5976.76  S&P500
3 2024-11-07  5973.10  5947.21  5983.84  5947.21  S&P500
4 2024-11-06  5929.04  5864.89  5936.14  5864.89  S&P500

In [ ]:
combined_df = pd.concat([nifty_df1, sp500_df, gold_df], ignore_index=True)
combined_df = combined_df[['Ticker', 'Date', 'Price', 'Open', 'High', 'Low']]

print("Combined DataFrame with desired column order:")
display(combined_df.head())
display(combined_df.tail())

Combined DataFrame with desired column order:


Ticker       Date      Price      Open      High       Low
0  Nifty-50 2005-01-03  48.732719  2,080.00  2,118.60  2,080.00
1  Nifty-50 2005-01-04  48.367628  2,116.95  2,120.15  2,100.55
2  Nifty-50 2005-01-05  46.685045  2,103.75  2,105.10  1,990.15
3  Nifty-50 2005-01-06  45.546439  2,031.55  2,035.65  1,984.25
4  Nifty-50 2005-01-07  46.068571  1,998.25  2,021.45  1,992.55

Ticker       Date      Price       Open       High        Low
16046   Gold 2023-09-01  62.642468  62.626393  63.050782  62.446348
16047   Gold 2023-08-31  62.597457   62.75178  62.873953  62.584597
16048   Gold 2023-08-30  62.822512  62.623177  62.951115  62.494575
16049   Gold 2023-08-29  62.256659  61.739033   62.27595  61.549343
16050   Gold 2023-08-28  61.681161  61.575064  61.922292  61.481827

# Task
Summarize the calculated historical returns for each ticker ('Nifty-50', 'S&P500', 'Gold') over 1-year, 3-year, 5-year, 10-year, and 15-year periods. The returns should be based on the 'Price' column in the `combined_df` and presented in a new DataFrame. Before calculating returns, ensure that the 'Open', 'High', and 'Low' columns of the `Nifty-50` data in `combined_df` are converted to a numeric type, handling any commas present in the string values.

## Calculate Returns for Each Ticker

### Subtask:
Calculate historical returns for each ticker ('Nifty-50', 'S&P500', 'Gold') over 1-year, 3-year, 5-year, 10-year, and 15-year periods using the 'Price' column in `combined_df`, presenting the results in a new DataFrame. Ensure 'Open', 'High', and 'Low' columns for Nifty-50 data are numeric first.


In [ ]:
nifty_rows = combined_df['Ticker'] == 'Nifty-50'
for col in ['Open', 'High', 'Low']:
    combined_df.loc[nifty_rows, col] = combined_df.loc[nifty_rows, col].astype(str).str.replace(',', '', regex=False).astype(float)

combined_df['Date'] = pd.to_datetime(combined_df['Date'])
combined_df = combined_df.sort_values(by=['Ticker', 'Date']).reset_index(drop=True)

print("Converted 'Open', 'High', 'Low' for Nifty-50 to numeric and ensured Date column is datetime, then sorted the DataFrame:")
print(combined_df[combined_df['Ticker'] == 'Nifty-50'].head())

Converted 'Open', 'High', 'Low' for Nifty-50 to numeric and ensured Date column is datetime, then sorted the DataFrame:
        Ticker       Date      Price     Open     High      Low
5589  Nifty-50 2005-01-03  48.732719   2080.0   2118.6   2080.0
5590  Nifty-50 2005-01-04  48.367628  2116.95  2120.15  2100.55
5591  Nifty-50 2005-01-05  46.685045  2103.75   2105.1  1990.15
5592  Nifty-50 2005-01-06  45.546439  2031.55  2035.65  1984.25
5593  Nifty-50 2005-01-07  46.068571  1998.25  2021.45  1992.55


In [ ]:
returns_data = []
time_periods = [1, 3, 5, 10, 15, 20]

for ticker_name, group in combined_df.groupby('Ticker'):
    group = group.sort_values(by='Date')
    latest_date = group['Date'].max()
    latest_price = group.loc[group['Date'] == latest_date, 'Price'].iloc[0]

    for years in time_periods:
        target_start_date = latest_date - pd.DateOffset(years=years)
        # Find the earliest date in the group that is >= target_start_date
        start_period_data = group[group['Date'] >= target_start_date]

        if not start_period_data.empty:
            # Get the price corresponding to the earliest date in the filtered data
            start_period_price = start_period_data.sort_values(by='Date')['Price'].iloc[0]
            if start_period_price != 0: # Avoid division by zero
                returns = ((latest_price - start_period_price) / start_period_price) * 100
            else:
                returns = None # Handle cases where start_period_price is zero
        else:
            returns = None # Not enough data for the period

        returns_data.append({
            'Ticker': ticker_name,
            'Period': f"{years}-Year",
            'Return (%)': returns
        })

historical_returns_df = pd.DataFrame(returns_data)

print("Historical Returns for Each Ticker:")
historical_returns_df.head(len(time_periods) * len(combined_df['Ticker'].unique()))

Historical Returns for Each Ticker:


Ticker   Period  Return (%)
0       Gold   1-Year   58.500729
1       Gold   3-Year  134.943166
2       Gold   5-Year  130.771744
3       Gold  10-Year  293.608918
4       Gold  15-Year  204.221712
5       Gold  20-Year  699.170124
6   Nifty-50   1-Year   -1.097542
7   Nifty-50   3-Year   27.885455
8   Nifty-50   5-Year   56.425832
9   Nifty-50  10-Year  152.080779
10  Nifty-50  15-Year  124.739082
11  Nifty-50  20-Year  380.307541
12    S&P500   1-Year   13.170325
13    S&P500   3-Year   74.106975
14    S&P500   5-Year   86.507025
15    S&P500  10-Year  234.539461
16    S&P500  15-Year  455.557989
17    S&P500  20-Year  443.912180

# Task
Prepare `gold_df` and `usd_inr_df` for correlation analysis by ensuring their 'Date' columns are set as the index, sorting both DataFrames by date, and selecting only the 'Price' column for each.

## Prepare Data for Correlation Analysis

### Subtask:
Ensure the 'Date' columns in `gold_df` and `usd_inr_df` are properly formatted and set as index, and select the relevant 'Price' columns for analysis. Sort both dataframes by date.


**Reasoning**:
To prepare `gold_df` for correlation analysis, I will set the 'Date' column as its index, sort the index, and then select only the 'Price' column. For `usd_inr_df`, I will select only its 'Price' column, as it has already been processed with the 'Date' column as its index and sorted.



In [ ]:
gold_df = gold_df.set_index('Date').sort_index()[['Price']]
usd_inr_df = usd_inr_df[['Price']]

print("Prepared gold_df:")
print(gold_df.head())
print("\nPrepared usd_inr_df:")
print(usd_inr_df.head())

Prepared gold_df:
                Price
Date                 
2005-01-03  13.815165
2005-01-04  13.799090
2005-01-05  13.738004
2005-01-06  13.554745
2005-01-07  13.487228

Prepared usd_inr_df:
             Price
2005-01-03  43.400
2005-01-04  43.495
2005-01-05  43.530
2005-01-06  43.875
2005-01-07  43.750


## Merge Gold and USD/INR Data

### Subtask:
Merge `gold_df` and `usd_inr_df` on their 'Date' column to create a combined DataFrame for correlation analysis. This will align the gold prices with the corresponding USD/INR exchange rates.


**Reasoning**:
To merge `gold_df` and `usd_inr_df` for correlation analysis, I will perform an inner join on their 'Date' index, rename the 'Price' columns for clarity, and then display the head of the resulting DataFrame.



In [ ]:
merged_df = gold_df.merge(usd_inr_df, left_index=True, right_index=True, how='inner', suffixes=('_gold', '_usd_inr'))
merged_df = merged_df.rename(columns={'Price_gold': 'Gold Price', 'Price_usd_inr': 'USD/INR Price'})

print("Merged Gold and USD/INR Data for correlation analysis:")
print(merged_df.head())

Merged Gold and USD/INR Data for correlation analysis:
            Gold Price  USD/INR Price
2005-01-03   13.815165         43.400
2005-01-04   13.799090         43.495
2005-01-05   13.738004         43.530
2005-01-06   13.554745         43.875
2005-01-07   13.487228         43.750


## Calculate Daily Percentage Change

### Subtask:
Calculate the daily percentage change for both gold prices and USD/INR exchange rates. This will help identify periods of price movement and INR depreciation.


**Reasoning**:
To calculate the daily percentage change for 'Gold Price' and 'USD/INR Price', I will use the `pct_change()` method on these columns in `merged_df` and multiply by 100 to express them as percentages.



In [ ]:
merged_df['Gold Daily Change (%)'] = merged_df['Gold Price'].pct_change() * 100
merged_df['USD/INR Daily Change (%)'] = merged_df['USD/INR Price'].pct_change() * 100

print("Merged DataFrame with daily percentage changes:")
print(merged_df.head())

Merged DataFrame with daily percentage changes:
            Gold Price  USD/INR Price  Gold Daily Change (%)  \
2005-01-03   13.815165         43.400                    NaN   
2005-01-04   13.799090         43.495              -0.116360   
2005-01-05   13.738004         43.530              -0.442684   
2005-01-06   13.554745         43.875              -1.333957   
2005-01-07   13.487228         43.750              -0.498102   

            USD/INR Daily Change (%)  
2005-01-03                       NaN  
2005-01-04                  0.218894  
2005-01-05                  0.080469  
2005-01-06                  0.792557  
2005-01-07                 -0.284900  


## Calculate and Display Correlation

### Subtask:
Compute the correlation coefficient between the daily percentage change of gold prices and the daily percentage change of USD/INR. Display the correlation matrix or coefficient.


**Reasoning**:
To compute the correlation coefficient, I will select the 'Gold Daily Change (%)' and 'USD/INR Daily Change (%)' columns from `merged_df` and apply the `.corr()` method.



In [ ]:
correlation = merged_df['Gold Daily Change (%)'].corr(merged_df['USD/INR Daily Change (%)'])

print(f"Correlation between Gold Daily Change and USD/INR Daily Change: {correlation:.4f}")

Correlation between Gold Daily Change and USD/INR Daily Change: -0.1992


## Final Task

### Subtask:
Summarize the findings regarding the correlation between gold price and INR depreciation.


## Summary:

### Q&A
The correlation between the daily percentage change of gold prices and the daily percentage change of USD/INR exchange rates is approximately -0.1992. This indicates a weak negative linear relationship.

### Data Analysis Key Findings
*   Data preparation involved setting the 'Date' column as the index for both `gold_df` and `usd_inr_df`, sorting them by date, and selecting only the 'Price' column for each.
*   `gold_df` and `usd_inr_df` were successfully merged into a single DataFrame, `merged_df`, using an inner join on their date indices. The price columns were renamed to 'Gold Price' and 'USD/INR Price' respectively.
*   Daily percentage changes were calculated for both 'Gold Price' and 'USD/INR Price', creating new columns named 'Gold Daily Change (%)' and 'USD/INR Daily Change (%)'.
*   The correlation coefficient between 'Gold Daily Change (%)' and 'USD/INR Daily Change (%)' was found to be approximately -0.1992, indicating a weak inverse relationship.

### Insights or Next Steps
*   The weak negative correlation suggests that a depreciation in the INR (increase in USD/INR price) is only slightly associated with a decrease in gold prices, and vice versa. This relationship is not strong enough to be a primary driver for investment decisions based solely on daily changes.
*   Further analysis could involve exploring correlation over different timeframes (e.g., weekly, monthly) or incorporating other macroeconomic factors that might influence both gold prices and the USD/INR exchange rate to identify stronger relationships.
